pgsqltoolsservice/driver/types/psycopg_driver.py

```diff
@@ -0,0 +1,178 @@
+}
+
+# Recognized parameter keywords for postgres database connection
+# Source: https://www.postgresql.org/docs/9.6/static/libpq-connect.html#LIBPQ-PARAMKEYWORDS
+PG_CONNECTION_PARAM_KEYWORDS = [
+    'host', 'hostaddr', 'port', 'dbname', 'user', 'password', 'passfile', 'connect_timeout',
+    'client_encoding', 'options', 'application_name', 'fallback_application_name', 'keepalives',
+    'keepalives_idle', 'keepalives_interval', 'keepalives_count', 'tty', 'sslmode', 'requiressl',
+    'sslcompression', 'sslcert', 'sslkey', 'sslrootcert', 'sslcrl', 'requirepeer', 'krbsrvname',
+    'gsslib', 'service', 'target_session_attrs'
+]
+
+
+class PsycopgConnection(ServerConnection):
+    """Wrapper for a psycopg2 connection that makes various properties easier to access"""
+
+    def __init__(self, conn_params):
+        """
+        Creates a new connection wrapper. Parses version string
+        :param conn_params: connection parameters dict
+        """
+        # Map the connection options to their psycopg2-specific options
+        connection_options = {PG_CONNECTION_OPTION_KEY_MAP.get(option, option): value for option, value in conn_params.items()
+        if option in PG_CONNECTION_PARAM_KEYWORDS}
+        
+        # Use the default database if one was not provided
+        if 'dbname' not in connection_options or not connection_options['dbname']:
+            connection_options['dbname'] = "postgres"
+
+        # Pass connection parameters as keyword arguments to the connection by unpacking the connection_options dict
+        self._conn = psycopg2.connect(**connection_options)
+
+        # Check that we connected successfully
+        assert type(self._conn) is connection
+        print("Connection to PostgreSQL server established!")
+
+        # Set autocommit mode so that users have control over transactions
+        self._conn.autocommit = True
+
+        # Get the DSN parameters for the connection as a dict
+        self._dsn_parameters = self._conn.get_dsn_parameters()
+
+        # Find the class of the database error this driver throws
+        self.database_error = psycopg2.DatabaseError
+
+        # Calculate the server version
+        version_string = str(self._conn.server_version)
+        self._version: Tuple[int, int, int] = (
+            int(version_string[:-4]),
+            int(version_string[-4:-2]),
+            int(version_string[-2:])
+        )
+
+    ###################### PROPERTIES ##################################
+    @property
+    def connection(self):
+        """The psycopg2 connection that this object wraps"""
+        return self._conn
+
+    @property
+    def autocommit(self) -> bool:
+        """Returns the current autocommit status for this connection"""
+        return self._conn.autocommit
+
+    @property
+    def dsn_parameters(self) -> Mapping[str, str]:
+        """DSN properties of the underlying connection"""
+        return self._dsn_parameters
+    
+    @property
+    def database_name(self):
```

Comment:
Shouldn't this be in base class ? Also we should define the return type.

Reply:
Right, my bad! I added it to both the base class and the pymysql driver class as well.